### PR TITLE
More robust Fish autocompletion installation

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun 

This extended autocompletion 
* uses the already downloaded autocompletion definition, as suggested by @allewun 
* uses a symlink to the downloaded autocompletion definition, so that it auto-updates as Git Town is updated
* doesn't overwrite existing Git autocompletion definition files

### CODE REVIEW

**Comment by charlierudolph** [comment]
While I agree with not overwriting existing Git autocompletion definition files, I already brought up that on the previous [pr](https://github.com/Originate/git-town/pull/358#discussion-diff-23203406) that symlinking depends on installation via homebrew which is not guaranteed. I'm okay with optimizing the experience for someone that installed via homebrew but we should keep the existing functionality for users who did not install via homebrew.

**Comment by allewun** [comment]
@charlierudolph Where does the symlink depend on homebrew here? The previous implementation you linked to did, but this PR seems ok to me.

File: features/step_definitions/folder_steps.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+end
+
+
+Given(/^I have no fish autocompletion folder$/) do
+  FileUtils.rm_rf File.expand_path('~/.config/fish/completions')
+end
+
+
+Given(/^I have an empty fish autocompletion folder$/) do
+  suppress do
```
Comment (allewun) [inline features/step_definitions/folder_steps.rb]: What's `suppress`? I can't find any docs on this.

File: src/git-town [hunk 2]
```diff
@@ -1,11 +1,29 @@
...
+  autocomplete_folder=$(pwd)
+  cd -
+
+  # Optionally create the autocompeletion folder for Fish shell
+  if [ ! -d ~/.config/fish/completions ]; then
+    run_command_outside_git "mkdir -p ~/.config/fish/completions"
+  fi
+
+  # Ensure there are no existing autocompletions for Fish shell
+  if [ -e ~/.config/fish/completions/git.fish ] || [ -h ~/.config/fish/completions/git.fish ]; then
```
Comment (allewun) [inline src/git-town]: Seems like the first check is redundant, per [this reference](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html)

File: features/step_definitions/run_steps.rb [hunk 3]
```diff
@@ -73,6 +73,9 @@
 
 
 Then(/^it runs the following shell commands/) do |expected_commands|
+  expected_commands.map_column! 'COMMAND' do |command|
+    ERB.new(command).result
```
Comment (allewun) [inline features/step_definitions/run_steps.rb]: Using erb here is cool

File: features/git-town/setup/install_fish_autocomplete.feature [hunk 4]
```diff
@@ -5,9 +6,36 @@ Feature: Installing Fish Shell autocomplete definitions
...
+  Scenario: with empty fish autocompletion folder
+    Given I have an empty fish autocompletion folder
+    When I run `git town install-fish-autocompletion`
+    Then it runs the following shell commands to install the Fish shell autocompletion
+      | COMMAND                                                                                   |
+      | ln -s <%= GIT_TOWN_DIRECTORY %>/autocomplete/git.fish ~/.config/fish/completions/git.fish |
+
+
+  Scenario: with an existing Git autocompletion file
+    Given I have a different Git autocompletion file
```
Comment (allewun) [inline features/git-town/setup/install_fish_autocomplete.feature]: "a different" -> "an existing"

File: features/git-town/setup/install_fish_autocomplete.feature [hunk 5]
```diff
@@ -5,9 +6,36 @@ Feature: Installing Fish Shell autocomplete definitions
...
+    When I run `git town install-fish-autocompletion`
+    Then it runs no shell commands
+    And I get the error "You already have an existing Fish shell autocompletion definition for Git"
+    And I get the error "Operation aborted"
+    And I still have my different Git autocompletion file
```
Comment (allewun) [inline features/git-town/setup/install_fish_autocomplete.feature]: "different" -> "original"

File: features/step_definitions/folder_steps.rb (hunk 1 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: This is part of Rails' [ActiveSupport library](http://guides.rubyonrails.org/active_support_core_extensions.html), which extends Ruby with even more awesomeness.

Here is the [documentation for suppress](http://www.rubydoc.info/gems/activesupport/4.2.0/Kernel:suppress)

Here is where we load it: https://github.com/Originate/git-town/blob/master/features/support/env.rb#L1

**Comment by kevgo** [comment]
Yeah, this implementation is independent of where Git Town is installed. It locates the currently runnable "git-town" command, and traverses relative to that.

File: features/step_definitions/folder_steps.rb (hunk 1 above)
Comment (charlierudolph) [inline features/step_definitions/folder_steps.rb]: Suppressing sounds like a bad idea to me. Can you just use `FileUtils.rm_f`?

File: src/git-town [hunk 6]
```diff
@@ -1,11 +1,29 @@
...
 
 # Installs the autocompletion definition for Fish shell
 function install_fish_autocompletion {
-  run_command "mkdir -p ~/.config/fish/completions"
-  run_command "curl -o ~/.config/fish/completions/git.fish https://raw.githubusercontent.com/Originate/git-town/master/autocomplete/git.fish"
+
+  # Determine the Git Town source code directory that contains the autocompletion data
+  cd "$base_dir/../autocomplete"
+  autocomplete_folder=$(pwd)
+  cd -
```
Comment (charlierudolph) [inline src/git-town]:
```shell
local autocomplete_folder=$( cd "$base_dir/../autocomplete" && pwd )
```

File: features/step_definitions/folder_steps.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+Given(/^I already have the Git autocompletion symlink$/) do
+  FileUtils.rm File.expand_path('~/.config/fish/completions/git.fish')
+  FileUtils.symlink 'foo',
+                    File.expand_path('~/.config/fish/completions/git.fish')
+end
+
+
+Given(/^I have a different Git autocompletion file$/) do
+  FileUtils.rm File.expand_path('~/.config/fish/completions/git.fish')
+  IO.write File.expand_path('~/.config/fish/completions/git.fish'),
```
Comment (charlierudolph) [inline features/step_definitions/folder_steps.rb]: You are repeating `File.expand_path('~/.config/fish/completions/git.fish')` a ton of times. Lets throw that in a constant in  `env.rb`

File: features/git-town/setup/install_fish_autocomplete.feature [hunk 8]
```diff
@@ -5,9 +6,36 @@ Feature: Installing Fish Shell autocomplete definitions
...
+
+
+  Scenario: with existing Git autocompletion symlink
+    Given I already have the Git autocompletion symlink
+    When I run `git town install-fish-autocompletion`
+    Then it runs no shell commands
+    And I get the error "You already have an existing Fish shell autocompletion definition for Git"
```
Comment (charlierudolph) [inline features/git-town/setup/install_fish_autocomplete.feature]: I think this message can be shortened.
```
Git autocompletion for Fish shell already exists.
```

File: src/git-town (hunk 2 above)
Comment (kevgo) [inline src/git-town]: Yes, I thought that too, but it breaks certain specs.

File: features/git-town/setup/install_fish_autocomplete.feature (hunk 4 above)
Comment (kevgo) [inline features/git-town/setup/install_fish_autocomplete.feature]: done

File: features/git-town/setup/install_fish_autocomplete.feature (hunk 5 above)
Comment (kevgo) [inline features/git-town/setup/install_fish_autocomplete.feature]: done

File: src/git-town (hunk 6 above)
Comment (kevgo) [inline src/git-town]: This is clever! Thanks! done

File: features/step_definitions/folder_steps.rb (hunk 7 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: Good idea. Done

File: features/git-town/setup/install_fish_autocomplete.feature (hunk 8 above)
Comment (kevgo) [inline features/git-town/setup/install_fish_autocomplete.feature]: done

File: features/step_definitions/folder_steps.rb (hunk 1 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: Excellent idea! done

**Comment by kevgo** [comment]
Excellent comments, guys!

**Comment by kevgo** [comment]
All comments addressed

File: features/step_definitions/folder_steps.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
+Given(/^I already have the Git autocompletion symlink$/) do
+  FileUtils.rm FISH_AUTOCOMPLETIONS_PATH
+  FileUtils.symlink 'foo',
+                    FISH_AUTOCOMPLETIONS_PATH
```
Comment (allewun) [inline features/step_definitions/folder_steps.rb]: this is short enough to be on 1 line now

File: features/step_definitions/folder_steps.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+end
+
+
+Given(/^I have an existing Git autocompletion file$/) do
+  FileUtils.rm FISH_AUTOCOMPLETIONS_PATH
+  IO.write FISH_AUTOCOMPLETIONS_PATH,
+           'existing Git autocompletion data'
```
Comment (allewun) [inline features/step_definitions/folder_steps.rb]: this too

**Comment by allewun** [comment]
Just had an idea.. instead of just ending with "Operation aborted" when an autocompletion file exists already, do you think something like "Would you like to add the Git Town autocompletion code to your existing Git autocompletion file?" makes sense? If they say yes, then append the contents of our git.fish to their git.fish.

Otherwise, this PR can be :ship:

File: features/step_definitions/folder_steps.rb (hunk 9 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: Good catch. Fixed

File: features/step_definitions/folder_steps.rb (hunk 10 above)
Comment (kevgo) [inline features/step_definitions/folder_steps.rb]: fixed

**Comment by kevgo** [comment]
@allewun I'd rather not mess with any existing configuration programmatically. Chances are high that this will cause duplication and/or not work as expected. And then Git Town is to  blame for that, and people will hate us.

**Comment by kevgo** [comment]
all comments addressed

**Comment by charlierudolph** [comment]
:ship:

**Pull request closed without merging**